### PR TITLE
feat: [PL-38977]: Add api to delete all users except the admins

### PR DESCRIPTION
**Pull request by sahibaCCM**
## Describe your changes
The API deletes all the NG users in an account accept for admins.
-> tested for created multiple admins and non admin users

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>

https://github.com/harness/harness-core/assets/112424209/b7a435b1-2710-4c29-bcf6-f417c1925cd4

  <summary>Build triggers</summary>

https://github.com/harness/harness-core/assets/112424209/f073bf11-f3cf-40ab-82c5-192152645d04

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)

### CODE REVIEW

**Comment by sahibaCCM** [comment]
trigger messagecheck

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java [hunk 1]
```diff
@@ -331,4 +329,13 @@ private void sleep() {
       log.warn("Thread Interrupted", ex);
     }
   }
+
+  public void cleanAccountForNG(String accountIdentifier) {
+    Collection<UserInfo> cgUsers = getCGUsers(accountIdentifier);
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java:334]: Get users from NG collection `userMenmbership`

**Comment by sahibaCCM** [comment]
trigger uts

**Review by tiwaritejas — CHANGES_REQUESTED** [review]
> Reviewed 3 of 10 files at r1, 2 of 4 files at r3, 3 of 6 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)* status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java` line 1126 at r6](https://reviewable.io/reviews/harness/harness-core/48743#-NXxGwvo7BKlTY2yASNG:-NXxGwvo7BKlTY2yASNH:b3uofd7) ([raw file](https://github.com/harness/harness-core/blob/f3a5d885658e3b54fe0e4c9fd3be1f2673a9d0c1/123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java#L1126)):*
> ```Java
>     Criteria userMembershipCriteria = getCriteriaForFetchingChildScopesForUserIds(userIds, scope);
>     Page<UserMembership> userMemberships =
>         userMembershipRepository.findAll(userMembershipCriteria, Pageable.ofSize(userIds.size()));
> ```

A single user can have multiple entries in the userMemberships collection (one corresponding to each scope that they are a part of in that account). Hence. passing the page size as the size of the userIds will miss out some userMemberships. How about we make it a non-paginated call here? 
___
*[`400-rest/src/main/java/software/wings/service/impl/AccountServiceImpl.java` line 579 at r6](https://reviewable.io/reviews/harness/harness-core/48743#-NXxHSK64JM2yQQnUsmr:-NXxHSK64JM2yQQnUsms:b-ri6u72) ([raw file](https://github.com/harness/harness-core/blob/f3a5d885658e3b54fe0e4c9fd3be1f2673a9d0c1/400-rest/src/main/java/software/wings/service/impl/AccountServiceImpl.java#L579)):*
> ```Java
>   public Boolean cleanUpNextGen(String accountId) {
>     Account account = get(accountId);
>     update(account);
> ```

Why is this update(account) call required here?

**Comment by sahibaCCM** [comment]
> 

-> fixing the pagable part
-> was setting nextgen-enabled as false; removed it as asked by shashank but update part forgot to remove. Fixing

**Comment by sahibaCCM** [comment]
> > Reviewed 3 of 10 files at r1, 2 of 4 files at r3, 3 of 6 files at r5, 3 of 3 files at r6, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)_ status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained.
> 
> _[`123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java` line 1126 at r6](https://reviewable.io/reviews/harness/harness-core/48743#-NXxGwvo7BKlTY2yASNG:-NXxGwvo7BKlTY2yASNH:b3uofd7) ([raw file](https://github.com/harness/harness-core/blob/f3a5d885658e3b54fe0e4c9fd3be1f2673a9d0c1/123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java#L1126)):_
> 
> > ```java
> >     Criteria userMembershipCriteria = getCriteriaForFetchingChildScopesForUserIds(userIds, scope);
> >     Page<UserMembership> userMemberships =
> >         userMembershipRepository.findAll(userMembershipCriteria, Pageable.ofSize(userIds.size()));
> > ```
> 
> A single user can have multiple entries in the userMemberships collection (one corresponding to each scope that they are a part of in that account). Hence. passing the page size as the size of the userIds will miss out some userMemberships. How about we make it a non-paginated call here?
> 
> _[`400-rest/src/main/java/software/wings/service/impl/AccountServiceImpl.java` line 579 at r6](https://reviewable.io/reviews/harness/harness-core/48743#-NXxHSK64JM2yQQnUsmr:-NXxHSK64JM2yQQnUsms:b-ri6u72) ([raw file](https://github.com/harness/harness-core/blob/f3a5d885658e3b54fe0e4c9fd3be1f2673a9d0c1/400-rest/src/main/java/software/wings/service/impl/AccountServiceImpl.java#L579)):_
> 
> > ```java
> >   public Boolean cleanUpNextGen(String accountId) {
> >     Account account = get(accountId);
> >     update(account);
> > ```
> 
> Why is this update(account) call required here?

-> fixing the pagable part
-> was setting nextgen-enabled as false; removed it as asked by shashank but update part forgot to remove. Fixing

**Comment by sahibaCCM** [comment]
trigger uts

**Review by tiwaritejas — DISMISSED** [review]
> Reviewed 2 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Comment by sahibaCCM** [comment]
trigger ut0

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/event/AccountSetupListener.java` line 104 at r8](https://reviewable.io/reviews/harness/harness-core/48743#-NXz8AEQ0aMZsOFCOg1d:-NXz8AEQ0aMZsOFCOg1e:b36fjjl) ([raw file](https://github.com/harness/harness-core/blob/a850489e006aa896be886d3ffd59d6cb653fda4e/120-ng-manager/src/main/java/io/harness/ng/core/event/AccountSetupListener.java#L104)):*
> ```Java
>     log.info(String.format("[AccountSetupListener]: Received account cleanup event for account %s",
>         accountEntityChangeDTO.getAccountId()));
>     AccountDTO account = CGRestUtils.getResponse(accountClient.getAccountDTO(accountEntityChangeDTO.getAccountId()));
> ```

not sure if I understand the need for this call. Either this call is not needed or we should add check (account != null) , thanks



_Code quote:_
```Java
 AccountDTO account = CGRestUtils.getResponse(accountClient.getAccountDTO(accountEntityChangeDTO.getAccountId()));
```

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java` line 333 at r8](https://reviewable.io/reviews/harness/harness-core/48743#-NXz8cr642_7_EDb7f6K:-NXz8cr642_7_EDb7f6L:b-x0osx) ([raw file](https://github.com/harness/harness-core/blob/a850489e006aa896be886d3ffd59d6cb653fda4e/120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java#L333)):*
> ```Java
>   }
> 
>   public void cleanAccountForNG(String accountIdentifier) {
> ```

nit: methodName is little misleading, it should be cleanUsersFromAccountForNg, thanks



_Code quote:_
```Java
public void cleanAccountForNG(String accountIdentifier) {
```

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java` line 337 at r8](https://reviewable.io/reviews/harness/harness-core/48743#-NXz8mjz86ywdzJiVQR5:-NXz8mjz86ywdzJiVQR6:b8hrznj) ([raw file](https://github.com/harness/harness-core/blob/a850489e006aa896be886d3ffd59d6cb653fda4e/120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java#L337)):*
> ```Java
>         Scope.builder().accountIdentifier(accountIdentifier).orgIdentifier(null).projectIdentifier(null).build();
>     List<String> ngUsers = ngUserService.listUserIds(scope);
>     log.info("ngUsers  {}", ngUsers);
> ```

minor: lets improve the log line here:
For account {}, users found in NG: {}



_Code quote:_
```Java
log.info("ngUsers  {}", ngUsers);
```

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java` line 341 at r8](https://reviewable.io/reviews/harness/harness-core/48743#-NXz9HdF26yPXiKrX_O-:-NXz9HdF26yPXiKrX_O0:b-8un1gh) ([raw file](https://github.com/harness/harness-core/blob/a850489e006aa896be886d3ffd59d6cb653fda4e/120-ng-manager/src/main/java/io/harness/ng/core/event/NGAccountSetupService.java#L341)):*
> ```Java
>                                    .filter(user -> !ngUserService.isAccountAdmin(user, accountIdentifier))
>                                    .collect(Collectors.toList());
>     log.info("ngNonAdmins  {}", ngNonAdmins);
> ```

minor: same as above:
For account {}, non admin users found in NG: {}



_Code quote:_
```Java
log.info("ngNonAdmins  {}", ngNonAdmins);
```

**Review by pbarapatre10 — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 1 of 6 files at r5, 1 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)* status: 8 file(s) to review. 1 of 2 approvals obtained.

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/test/java/io/harness/ng/core/event/NGAccountSetupServiceTest.java [hunk 2]
```diff
@@ -204,4 +205,17 @@ public void testDefaultProjectIsDisabledAndOnlyCGAdminsMigratedToNG() {
     assertThat(users.contains(FIRST_ADMIN_USER)).isTrue();
     assertThat(users.contains(SECOND_ADMIN_USER)).isTrue();
   }
+
+  @Test
+  @Owner(developers = SAHIBA)
+  @Category(UnitTests.class)
+  public void testCleanUpForNG() {
+    when(featureFlagService.isGlobalEnabled(FeatureName.CREATE_DEFAULT_PROJECT)).thenReturn(false);
+    when(featureFlagService.isNotEnabled(FeatureName.PL_DO_NOT_MIGRATE_NON_ADMIN_CG_USERS_TO_NG, ACCOUNT_ID))
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 120-ng-manager/src/test/java/io/harness/ng/core/event/NGAccountSetupServiceTest.java:214]: To create the users that we later are cleaning up

File: 120-ng-manager/src/test/java/io/harness/ng/core/event/NGAccountSetupServiceTest.java [hunk 3]
```diff
@@ -204,4 +205,17 @@ public void testDefaultProjectIsDisabledAndOnlyCGAdminsMigratedToNG() {
...
+        .thenReturn(false);
+    ngAccountSetupService.setupAccountForNG(ACCOUNT_ID);
+    ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 120-ng-manager/src/test/java/io/harness/ng/core/event/NGAccountSetupServiceTest.java:217]: I was checking in for admin creation will remove the line; not needed

File: 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java [hunk 4]
```diff
@@ -1115,4 +1116,45 @@ public UserMetadata updateUserMetadataInternal(UserMetadataDTO user) {
...
+    Scope scope =
+        Scope.builder().accountIdentifier(accountIdentifier).orgIdentifier(null).projectIdentifier(null).build();
+    Criteria userMembershipCriteria = getCriteriaForFetchingChildScopesForUserIds(userIds, scope);
+    List<UserMembership> userMemberships = new ArrayList<>();
+    Page<UserMembership> response = null;
+    int page = 0;
+    int size = 500;
+    do {
+      Pageable pageable = Pageable.ofSize(size).withPage(page);
+      response = userMembershipRepository.findAll(userMembershipCriteria, pageable);
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java:1131]: I'm using an existing function that is a paginated call; I can add another in UserMembershipRepositoryCustomImpl that can fetch all at once if that is needed.

File: 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java [hunk 5]
```diff
@@ -1115,4 +1116,45 @@ public UserMetadata updateUserMetadataInternal(UserMetadataDTO user) {
...
+
+      if (response != null && isNotEmpty(response.getContent())) {
+        userMemberships.addAll(response.getContent());
+      }
+      page++;
+    } while (response != null && isNotEmpty(response.getContent()));
+
+    Failsafe.with(transactionRetryPolicy).get(() -> transactionTemplate.execute(status -> {
+      log.info("Deleting accounts non admin only", userMemberships);
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java:1140]: This logs only this
"deleting accounts non admin only [entityType=account, action=ngUserCleanUp, producer=Manager ]"
Will move it post delete call.

File: 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java [hunk 6]
```diff
@@ -1115,4 +1116,45 @@ public UserMetadata updateUserMetadataInternal(UserMetadataDTO user) {
...
+      userMembershipRepository.deleteAll(userMemberships);
+      for (UserMembership user : userMemberships) {
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java:1142]: okay

**Comment by sahibaCCM** [comment]
trigger uts

**Review by pbarapatre10 — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)* status: Some of the checks are too old. 11 file(s) to review. 1 change requests. 5. unresolved discussions. 1 of 2 approvals obtained.

File: 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java:1131]: Added a function to make it a non paginated call.

File: 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java [hunk 7]
```diff
@@ -1115,4 +1116,45 @@ public UserMetadata updateUserMetadataInternal(UserMetadataDTO user) {
...
+    int size = 500;
+    do {
+      Pageable pageable = Pageable.ofSize(size).withPage(page);
+      response = userMembershipRepository.findAll(userMembershipCriteria, pageable);
+
+      if (response != null && isNotEmpty(response.getContent())) {
+        userMemberships.addAll(response.getContent());
+      }
+      page++;
+    } while (response != null && isNotEmpty(response.getContent()));
```
[the remark this answers is not in the document]
Comment (sahibaCCM) [reply 123-ng-core-user/src/main/java/io/harness/ng/core/user/service/impl/NgUserServiceImpl.java:1137]: getAllUserMemberships not needed as added a paginated call

**Comment by sahibaCCM** [comment]
trigger uts

**Review by meenakshi-harness — APPROVED** [review]

**Review by pbarapatre10 — APPROVED** [review]
> Reviewed 3 of 10 files at r1, 1 of 4 files at r3, 1 of 6 files at r5, 1 of 5 files at r9, 4 of 6 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48743)* status: 8 file(s) to review. 6. unresolved discussions.

**Pull request merged**